contracts/marketplace/src/lib.rs

```diff
@@ -0,0 +1,215 @@
+        assets.get(asset)
+    }
+
+    /// Allow sellers to list assets by specifying the seller's address, the asset's address, and the asset's price.
+    pub fn create_listing(env: Env, seller: Address, asset: Address, price: i128) {
+        if price <= 0 {
+            panic_with_error!(&env, Error::InvalidAssetPrice);
+        }
+        seller.require_auth();
+        // TODO: Check if asset is owned by seller
+        let storage = env.storage().persistent();
+        let mut assets = storage.get(&DataKey::Assets).unwrap_or(Map::new(&env));
+        assets.set(
+            asset,
+            Asset {
+                owner: seller,
+                price,
+                listed: true,
+            },
+        );
+        storage.set(&DataKey::Assets, &assets);
+    }
+
+    /// Enable buyers to purchase assets by providing the buyer's address, the asset's address, and the agreed-upon price.
+    pub fn buy_listing(env: Env, buyer: Address, asset: Address, price: i128) {
+        if price <= 0 {
+            panic_with_error!(&env, Error::InvalidAssetPrice);
+        }
+        buyer.require_auth();
+        let storage = env.storage().persistent();
+        let token = storage.get(&DataKey::Token).unwrap();
+        let mut assets: Map<Address, Asset> = storage.get(&DataKey::Assets).unwrap();
+        let Asset {
+            owner: seller,
+            price: set_price,
+            listed,
+        } = assets.get(asset.clone()).unwrap();
+        if !listed {
+            panic_with_error!(&env, Error::AssetNotListed);
+        }
+        if price != set_price {
+            panic_with_error!(&env, Error::InvalidAssetPrice);
+        }
+        let token = token::Client::new(&env, &token);
+        if token.balance(&buyer) < price {
+            panic_with_error!(&env, Error::BalanceTooLow);
+        }
+        token.transfer(&buyer, &seller, &price);
+        assets.set(
+            asset,
+            Asset {
+                owner: buyer,
+                price,
+                listed: false,
+            },
+        );
+        storage.set(&DataKey::Assets, &assets);
+    }
+
+    /// Permit sellers to update the price of a listed asset,
+    /// ensuring they provide the correct seller and asset addresses, as well as the old and new prices.
+    pub fn update_listing(
+        env: Env,
+        seller: Address,
+        asset: Address,
+        old_price: i128,
+        new_price: i128,
+        listed: bool,
+    ) {
+        if new_price <= 0 {
+            panic_with_error!(&env, Error::InvalidAssetPrice);
+        }
+        seller.require_auth();
+        let storage = env.storage().persistent();
+        let mut assets: Map<Address, Asset> = storage.get(&DataKey::Assets).unwrap();
+        let Asset {
+            owner: set_seller,
+            price: set_price,
+            ..
+        } = assets.get(asset.clone()).unwrap();
+        assert_eq!(set_seller, seller);
```

Comment:
Why `assert_eq` is used here and the prices below are compared using `if` and `panic_with_error`?